SNAP-1792: Display snappy members logs on Snappy Pulse UI

snappy-sachin opened the pull request:
## Changes proposed in this pull request

- Code changes to retrieve members log file path.

## Patch testing

- Manually Tested.

## ReleaseNotes changes

- None

## Other PRs 
SnappyData: https://github.com/SnappyDataInc/snappydata/pull/697
Spark: https://github.com/SnappyDataInc/spark/pull/58

snappy-sachin committed:
SNAP-1792: Display snappy members logs on Snappy Pulse UI
Changes:
  - Code changes to retrieve members log file path.

snappy-sachin committed:
Merge branch 'snappy/master' into SNAP-1792

snappy-sachin committed:
SNAP-1792: Display snappy members logs on Snappy Pulse UI
Changes:
 - Adding MemberLogsMessage to fetch members logs and log details from members local file system.

snappy-sachin committed:
SNAP-1792: Display snappy members logs on Snappy Pulse UI
Changes:
 - Adding LogFileComparator to sort log files based on log file names and rolling counter.

rishitesh commented:
Minor comments . LGTM.

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/sql/execute/MemberLogsMessage.java:
```diff
@@ -0,0 +1,392 @@
+  private Long offset;
+  //private int byteLength;
+  private Long byteLength;
+
```
why Long objects and not long primitive?

snappy-sachin replied:
offset could be null so using Long type (Primitive data types cannot be null). 
byteLength can be changed to long.

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/sql/execute/MemberLogsMessage.java:
```diff
@@ -0,0 +1,392 @@
+
+  /** Get the part of the log files given the offset and desired length of bytes */
+  private Map<String, Object> getLog(Long offset, Long byteLength) {
+
```
why not long primitives?

snappy-sachin replied:
Primitive data types cannot be null and offset could be null. Hence using Long.

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/sql/execute/MemberLogsMessage.java:
```diff
@@ -0,0 +1,392 @@
+      // set offset if not specified
+      if(offset == null){
+        offset = totalLength - byteLength;
+      }
```
What's the case for null offset and negative offset? Looks like this could use long primitives and negative to set it like above.

snappy-sachin replied:
Use Case 1: offset could be null:
offset and byteLength are not mandatory in http request as user/view may just need to see latest log entries with default size. Therefore offset could be null/unset.

Use Case 2: offset could be zero/number:
If http request sets offset as zero that means user/view needs to see log entries from the start of logs (i.e. when logging started).
If http request sets offset as some number that means user/view needs to see log entries from the given offset in bytes.

There is no negative offset use case here.

sumwale replied:
@snappy-sachin But the code below this is checking for (offset < 0) and handling it specially? What I meant is that negative offset could be used to indicate case 1 "no offset provided in http request".

snappy-sachin replied:
I have added negative offset check for safe side because caller may send negative offset index which effectively be treated as zero only.
case 1 (null case) is different than zero offset. null offset basically means use default offset (which is total length less default / passed desired length of logs).

sumwale replied:
Caller sending negative offset should throw an error or something no? What happens in this code if the offset requested lies beyond the size?

snappy-sachin replied:
if requested offset is beyond the  _total length of logs_ then _total length of logs_ is set as offset.

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/sql/execute/MemberLogsMessage.java:
```diff
@@ -0,0 +1,392 @@
+
+    Map<String, Object> logsMap = new HashMap<String, Object>();
+
+    String logFile = getLogFile();
```
You could pass this from execute() which has already determined this instead of determining again.

snappy-sachin replied:
Now passing log file name and directory as arguments to getLog method call.

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/sql/execute/MemberLogsMessage.java:
```diff
@@ -0,0 +1,392 @@
+
+  @Override
+  protected GfxdFunctionMessage clone() {
+    return null;
```
This can cause trouble. Provide a simple shallow clone or call a copy constructor which assigns the fields.

snappy-sachin replied:
Resolved. Its was missed.

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/sql/execute/MemberStatisticsMessage.java:
```diff
@@ -179,6 +184,21 @@ private String getUserDir(){
+      }
+    }
+    return "";
+  }
```
Same method is also in the MemberLogsMessage class above. Refactor into a common method.

snappy-sachin replied:
Removed getUserDir and getLogFile methods from MemberLogMessage. Setting them in instance variables while creating new MemberLogMessage object.

sumwale reviewed (commented):
Some minor comments and questions. Haven't looked at the logic of offsets yet.

snappy-sachin committed:
Merge branch 'snappy/master' into SNAP-1792

snappy-sachin committed:
SNAP-1545 feedback: Making use of Use ProcessUtils.identifyPid() to get members process id.

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/sql/execute/MemberStatisticsMessage.java:
```diff
@@ -179,9 +186,29 @@ private String getUserDir(){
+      processId = Integer.toString(ProcessUtils.identifyPid());
+    } catch (PidUnavailableException e) {
+      logger.warn("Process ID NOT AVAILABLE..", e);
+    }
```
@snappy-sachin Use NativeCalls.getInstance().getProcessId(). That works for all known platforms using native JNA and is much more reliable that this custom parsing so should be preferred everywhere. @rishitesh

snappy-sachin replied:
Yes, now using _NativeCalls.getInstance().getProcessId()_

rishitesh commented:
@sumwale @snappy-sachin I think ProcessUtils is used by launcher processes as well. If that breaks, I guess we have bigger problems of wrong PID going to pid files. The RunTimeMXbean also uses getProcessId() , but the parsing with "@" might cause problem with different flavors of java. Not sure though.

sumwale commented:
@rishitesh @snappy-sachin This parsing thing is undocumented and fragile. Doesn't work on Windows and likely other platforms (need to check on Mac), doesn't work on many other JDKs etc. All code that depends on processId being reliable should use NativeCalls.getInstance().getProcessId(). All of GemXD/Gem core that requires it was updated to use it at some point, but looks like mbeans have been using something else.

snappy-sachin committed:
SNAP-1792: Display snappy members logs on Snappy Pulse UI
Changes:
 - Adding LogFileUtils class which reads given log file content with specified length of logs content and offset position in log file.

snappy-sachin committed:
SNAP-1792: Display snappy members logs on Snappy Pulse UI
Changes:
 - Moved logic of reading logs from file into separate utility class.

snappy-sachin committed:
SNAP-1792: Display snappy members logs on Snappy Pulse UI
Changes:
 - Making use of NativeCalls.getInstance().getProcessId() to get process id.

snappy-sachin committed:
SNAP-1792: Display snappy members logs on Snappy Pulse UI
Changes:
 - Removing getUsedDir and getLogFile methods. Instead passing same as arguments.

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/sql/execute/MemberLogsMessage.java:
```diff
@@ -0,0 +1,182 @@
+  }
+
+  private MemberLogsMessage(final MemberLogsMessage other) {
+    super(other);
```
Also assign the extra fields: memberId, offset etc

snappy-sachin replied:
Missed it. I will do it.

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/sql/execute/MemberStatisticsMessage.java:
```diff
@@ -37,9 +41,12 @@
 
 public class MemberStatisticsMessage extends MemberExecutorMessage {
-
+  private Logger logger = Logger.getLogger(MemberStatisticsMessage.class);
```
not used?

snappy-sachin replied:
Removed unused variable..

snappy-sachin commented:
All feedback comments have been taken care.

Merging..

sumwale reviewed (commented):
@snappy-sachin see couple more minor comments
